Add an implementation of the algorithm for graph clusterisation

IlyaSuponev opened the pull request:
Добавлена реализация алгоритма [кластеризации](https://en.wikipedia.org/wiki/Cluster_analysis):

* [Louvain algorithm](https://arxiv.org/pdf/0803.0476).

IlyaSuponev committed:
feat: add set partition implementation

IlyaSuponev committed:
feat: add weighted adjacency matrix implementation

IlyaSuponev committed:
feat: add utils to compare double numbers by equality

IlyaSuponev committed:
feat: add method to search clusters at graphs

Add graph's modularity evaluator for Louvain method for searching clusters.

IlyaSuponev committed:
refactor: change codestyle by linter

IlyaSuponev committed:
feat: update clustering method

Update louvain clustering method by caсhing some sub-results of vertex nesting in modularity value.

IlyaSuponev committed:
fix: update code style and update tests cases

RodionovMaxim05 commented on graphs-lab/src/main/kotlin/graphs_lab/algs/clustering/LouvainClusteringMethod.kt:
```diff
@@ -0,0 +1,72 @@
+import graphs_lab.core.edges.Edge
+import graphs_lab.core.graphs.Graph
+
+/**
+ * Louvain clustering method implementation.
+ *
+ * See more: [Wiki](https://en.wikipedia.org/wiki/Louvain_method).
+ *
+ * @param I the type of the vertex ID
+ * @param E the type of the edge
+ * @param graph which need to search clusters
+ * @return pair of clustering partition and result modularity value
+ */
+fun <I, E : Edge<I>> louvainClusteringMethod(graph: Graph<I, E>): Pair<SetPartition<I>, Double> {
+	val partition = SetPartition(graph.idVertices)
+	val modularity: Double = louvainClusteringMethod(graph, partition)
+	return Pair(partition, modularity)
+}
+
+/**
+ * Louvain clustering method implementation.
+ *
+ * See more: [Wiki](https://en.wikipedia.org/wiki/Louvain_method).
+ *
+ * @param I the type of the vertex ID
+ * @param E the type of the edge
+ * @param graph which need to search clusters
+ * @param partition start clustering partition for [graph]
+ * @return result modularity value for [partition]
+ */
```
Может быть, стоит добавить в документацию второй функции, что пользователь может сам добавить начальное разбиение, чтобы он понимал, чем они отличаются из описания?

IlyaSuponev replied:
Верное замечание, добавлю.

RodionovMaxim05 commented on graphs-lab/src/main/kotlin/graphs_lab/algs/utils/WeightedAdjacencyMatrix.kt:
```diff
@@ -0,0 +1,97 @@
+		updateMatrix(graph)
+	}
+
+	/**
+	 * Get edge weight from graph.
+	 *
+	 * @param idSource id of edge's source vertex
+	 * @param idTarget id of edge's target vertex
+	 * @return `0.0` if edge not found else its weight, by result of [getEdgeWeight] method
+	 */
+	fun getEdgeWeight(idSource: I, idTarget: I): Double {
+		return getEdgeWeight(idSource, idTarget, 0.0)
+	}
+
+	/**
+	 * Get edge weight from graph.
+	 *
+	 * @param idSource id of edge's source vertex
+	 * @param idTarget id of edge's target vertex
+	 * @param defaultWeight value of not found edge which source is [idSource] and target is [idTarget]
+	 * @return [defaultWeight] if edge not found else its weight, by result of [getEdgeWeight] method
+	 */
+	fun getEdgeWeight(idSource: I, idTarget: I, defaultWeight: Double): Double {
+		return matrix.getOrDefault(idSource, mutableMapOf()).getOrDefault(idTarget, defaultWeight)
+	}
```
Тут то же самое, можно даже название второй функции изменить на 'getEdgeWeightOrDefault'

IlyaSuponev replied:
Переименную функцию, но здесь указано в доках что за параметр `defaultWeight` и когда он возвращается.

RodionovMaxim05 commented on graphs-lab/src/test/kotlin/graphs_lab/algs/utils/TestUtils.kt:
```diff
@@ -65,4 +65,16 @@ class TestUtils {
 		Assertions.assertEquals(checkAndGetSecond(pair), 1)
 	}
+
+	@Test
+	@DisplayName("check is double equals if its equals")
+	fun testDoubleEqualsCase1() {
+		Assertions.assertTrue(doubleEquality(0.0000001, 0.0000001))
+	}
```
Как будто, для того, чтобы точно протестировать функцию 'doubleEquality' нужно вместо одного из чисел давать результат арифметического действия.

IlyaSuponev replied:
Согласен, дополню тесты

IlyaSuponev committed:
refactor: update code style and test cases

Change method name on WeightedAdjacencyMatrix class from getEdgeWeight to getEdgeWeightOrDefault. Update KDos for louvainClusteringMethod with custom partition. Add tests for Utils:doubleEquality by math operation.

IlyaSuponev committed:
fix: backup name of util method to get edge weight

RodionovMaxim05 reviewed (approved):
Крутяк
